docs/isl-2-0/spec.md

```diff
@@ -0,0 +1,673 @@
+
+Generally speaking, constraints must reject null values as invalid.
+For example, the precision and scale constraints must reject a null value, as `null` does not have a precision or scale to evaluate.
+The `fields` constraint must reject `null.struct`, as `null.struct` doesn't have any fields.
+Similar reasoning applies to the expected handling of null values by most constraints.
+The `contains`, `annotations`, `valid_values`, and type-algebra constraints are exceptions to this, as these constraints may be defined such that a null value is valid.
+
+### Ranges
+
+{% include grammar-element.md productions="range_int,range_number,range_timestamp,range_timestamp_precision,number,exclusivity" %}
+
+Some constraints can be defined by a range.
+A range is represented by a list annotated with `range`, and containing two values, in order: the minimum and maximum ends of the range.
+The default behavior is for both ends of the range to be *inclusive*; 
+if *exclusive* behavior is desired, the minimum or maximum (or both) values shall be annotated with `exclusive`.
+If the minimum or maximum end of a range is to be unspecified, this shall be represented by the symbols `min` or `max`, respectively; 
+the `exclusive`annotation is not applicable when the symbols `min` or `max` are specified.
+A range may not contain both `min` and `max`.
+
+All ranges have a type.
+The type of the range is the same as that of the minimum and/or maximum values specified in the range list. 
+If both a minimum and maximum values are specified (i.e. `min` and `max` are not used), then both of those values must be of the same type. 
+(For example, `range::[1995-12-06T, 55.4]` mixes values of the timestamp and number types, and therefore is not a valid range.)
+
+
+## all_of
+
+{% include grammar-element.md productions="all_of" %}
+
+Value must be valid for all the types listed. The list of types must not be empty.
+
+{% include example.md title="`all_of` constraint" markdown="
+```ion
+type::{
+  name: Duck,
+  all_of: [
+    WalksLikeADuck,
+    QuacksLikeADuck,
+  ]
+}
+```
+" %}
+
+## annotations
+
+{% include grammar-element.md productions="annotations,annotations_modifier" %}
+
+Restricts the allowed annotations on a value.
+Applicable to all types except `document` (since documents cannot be annotated).
+This constraint has two available syntaxes.
+The "Standard" syntax is very expressive, but can be verbose for simple use cases.
+The "Simple" syntax is based on the Ion Schema 1.0 `annotations` syntax, and is concise, but only supports a limited set of use cases.
+
+### Standard Syntax
+
+The `annotations` constraint specifies the type and/or constraints for all annotations of a value.
+[The annotations themselves are symbols _tokens_](https://amzn.github.io/ion-docs/docs/spec.html#annot) but will be represented as a non-null list of non-null, un-annotated symbol _values_ for the purpose of validation.
```

Comment:
Phrasing suggestion:
```suggestion
[The annotations themselves are symbols _tokens_](https://amzn.github.io/ion-docs/docs/spec.html#annot) but will be considered a non-null list of non-null, un-annotated symbol _values_ for the purpose of validation.
```
Subjectively, "will be represented as" feels like an implementation detail.